bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -132,45 +132,51 @@ void ClientsManager::RequestsInfo::logAllPendingRequestsExceedingThreshold(const
 
 /*************************** Class ClientsManager::RepliesInfo ***************************/
 
-void ClientsManager::RepliesInfo::deleteOldestReplyIfNeededSafe(NodeIdType clientId, uint16_t maxNumOfReqsPerClient) {
-  Time earliestTime = MaxTime;
-  ReqId earliestReplyId = 0;
-  if (repliesMap_.size() < maxNumOfReqsPerClient) return;
+void ClientsManager::RepliesInfo::deleteReplyIfNeededSafe(NodeIdType clientId,
+                                                          ReqId reqSeqNum,
+                                                          uint16_t maxNumOfReqsPerClient,
+                                                          uint16_t reqOffset) {
+  ReqId oldReplyId = 0;
   if (repliesMap_.size() > maxNumOfReqsPerClient)
     LOG_FATAL(CL_MNGR,
               "More than maxNumOfReqsPerClient_ items in repliesInfo"
                   << KVLOG(repliesMap_.size(), maxNumOfReqsPerClient, clientId));
+  const lock_guard<mutex> lock(repliesMapMutex_);
+
+  // TODO: create another/additional data structure for efficiency (map offset to seqNum)
   for (auto& reply : repliesMap_) {
-    if (earliestTime > reply.second) {
-      earliestReplyId = reply.first;
-      earliestTime = reply.second;
+    if (reply.second == reqOffset) {
+      if (reply.first > reqSeqNum) LOG_WARN(CL_MNGR, "A newer reply was already saved");
```

Comment:
1. We shouldn't get to this warning. If we do, it means that something's wrong.
2. The offset (the index of request in batch) determines the reserved page in which the reply will be saved. So when we have new reply to be saved, we should remove from map the old reply that is now saved in the same page and will be overridden by the new one.
There cannot be such case as you described. In case the client retransmits a request, s/he re-sends the whole batch, so the request keeps its index in the batch.
3. We expect that the saved reply would be older than the one we are about to save. That's why we have this warning - for the unexpected case where the saved reply is newer than the one we have now.

Reply:
1. We don't expect this warning to happen. request should arrive in ascending order
2. There is 1:1 mapping between page id and offset (index in batch).
    A client request will always have the same index in batch because since a retransmit of a batch requests would always be a retransmit of the whole batch.
3. We expect to see an older reply in map